### ISSUE TITLE
Should type_annotate_public_apis lint final field?

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#1319]
`type_annotate_public_apis` lints the following code:

```dart
class A {
  final a = ''; // LINT
}
```

As `a` is final it's always a `String`. Should the lint be discarded in this case of final vars?

**Comment by bwilkerson** [comment]
Good question. Type inference might change this and several other style-oriented lints.

**Comment by kevmoo** [comment]
Let me enthusiastically agree with this!

If the type can be inferred (and it's a public type) then it shouldn't cause a lint!

**Comment by kevmoo** [comment]
CC @pq

**Comment by pq** [comment]
It's interesting to note that the referenced [style guideline](
https://dart.dev/guides/language/effective-dart/design#prefer-type-annotating-public-fields-and-top-level-variables-if-the-type-isnt-obvious) does allow for some wiggle when it comes to "obvious" types.  I'm definitely open to refinements.

What does `dartdoc` do in these cases?  (And does that matter?)

/cc @jcollins-g

**Comment by kevmoo** [comment]
dartdoc "does the right thing" 😄 

https://pub.dev/documentation/source_gen/latest/source_gen/defaultFileHeader-constant.html

**Comment by mr-mmmmore** [comment]
Will something be done about this? At the moment we still have to write

```dart
Class A {
    final MyClass myProperty = MyClass();
}
```

to comply with the `type_annotate_public_apis` rule.

But effective dart says [PREFER type annotating public fields and top-level variables **if the type isn’t obvious**](https://dart.dev/guides/language/effective-dart/design#prefer-type-annotating-public-fields-and-top-level-variables-if-the-type-isnt-obvious):

> In some cases, though, the type is so obvious that writing it is pointless:
> 
> ```dart
> const screenWidth = 640; // Inferred as int.
> ```
> 
> “Obvious” isn’t precisely defined, but these are all good candidates:
> 
> - Literals.
> - **Constructor invocations.**
> - References to other constants that are explicitly typed.
> - Simple expressions on numbers and strings.
> - Factory methods like int.parse(), Future.wait(), etc. that readers are expected to be familiar with.

**Comment by pq** [comment]
Thanks for the nudge @mr-mmmmore.  I think we've reached consensus that we want to support this and I'll put it  on my short-list to  take a look at updating the rule.

**Comment by mr-mmmmore** [comment]
Cool! Thank you @pq
